### PR TITLE
Fixed velocity support

### PR DESCRIPTION
**Pull request by Alvsch**
Fixed the support for velocity proxies

### CODE REVIEW

File: pumpkin-protocol/src/bytebuf/mod.rs [hunk 1]
```diff
@@ -370,6 +370,7 @@ impl ByteBuffer {
         if self.buffer.len() >= len {
             Ok(self.buffer.copy_to_bytes(len))
         } else {
+            dbg!("{:?} < {:?}", self.buffer.len(), len);
```
Comment (Snowiiii) [inline pumpkin-protocol/src/bytebuf/mod.rs:373]: Please remove that

File: pumpkin/src/client/mod.rs [hunk 2]
```diff
@@ -221,22 +220,24 @@ impl Client {
...
+            pumpkin_protocol::ConnectionState::Status => self.handle_status_packet(server, packet).unwrap(),
             // TODO: Check config if transfer is enabled
             pumpkin_protocol::ConnectionState::Login
             | pumpkin_protocol::ConnectionState::Transfer => {
-                self.handle_login_packet(server, packet).await
+                self.handle_login_packet(server, packet).await.unwrap()
             }
             pumpkin_protocol::ConnectionState::Config => {
-                self.handle_config_packet(server, packet).await
+                self.handle_config_packet(server, packet).await.unwrap()
```
Comment (Snowiiii) [inline pumpkin/src/client/mod.rs:233]: -_- We don't want unwraps

File: pumpkin/src/proxy/velocity.rs [hunk 3]
```diff
@@ -60,11 +60,36 @@ pub fn receive_plugin_response(
             ));
             return;
         }
-        // TODO: no unwrap
-        let addr: SocketAddr = buf.get_string().unwrap().parse().unwrap();
+        // TODO: dont default to localhost
+        let addr: SocketAddr = buf.get_string().unwrap().parse()
+            .unwrap_or_else(|_| ([127, 0, 0, 1], 25565).into());
```
Comment (Snowiiii) [inline pumpkin/src/proxy/velocity.rs:65]: you probably should get the current address then `BASIC_CONFIG.server_address

Comment (Alvsch) [reply pumpkin/src/proxy/velocity.rs:65]: that address is the default player address, it panics when you use localhost to connect to the server

Comment (Alvsch) [reply pumpkin/src/proxy/velocity.rs:65]: didn't come up with a better solution

Comment (Snowiiii) [reply pumpkin/src/proxy/velocity.rs:65]: then you should be able to use `client.address`.
Seems Paper does that as well: https://github.com/PaperMC/Paper/blob/master/patches/server/0731-Add-Velocity-IP-Forwarding-Support.patch#L202

Comment (Alvsch) [reply pumpkin/src/proxy/velocity.rs:65]: I was infact wrong about it being localhost, the address we get from velocity just doesn't include a port

Comment (Snowiiii) [reply pumpkin/src/proxy/velocity.rs]: > I was in fact wrong about it being localhost, the address we get from velocity just doesn't include a port edit: should we just default the port to 0?

looks like paper just grabs the current port from the players Socket Address, I think we should do that as well
image

**Comment by Snowiiii** [comment]
Looks good to me. 
Thanks you @Alvsch

**Pull request merged**